runtimes/js/src/pvalue.rs

```diff
@@ -0,0 +1,272 @@
+                        "Failed to detect whether given js is a date"
+                    )?;
+                    if is_date {
+                        let dt = JsDate::from_napi_value(env, napi_val)?;
+                        let secs = dt.value_of()?;
+                        let ts = chrono::Utc.timestamp_opt(secs as i64, 0);
+                        PValue::DateTime(ts.unwrap().fixed_offset())
+                    } else {
+                        PValue::Object(unsafe { PVals::from_napi_value(env, napi_val)?.0 })
+                    }
+                }
+            }
+            ValueType::BigInt => todo!(),
+            ValueType::Null => PValue::Null,
+            ValueType::Function => {
+                return Err(Error::new(
+                    Status::InvalidArg,
+                    "JS functions cannot be represented as a serde_json::Value".to_owned(),
+                ))
+            }
+            ValueType::Undefined => {
+                return Err(Error::new(
+                    Status::InvalidArg,
+                    "undefined cannot be represented as a serde_json::Value".to_owned(),
+                ))
+            }
+            ValueType::Symbol => {
+                return Err(Error::new(
+                    Status::InvalidArg,
+                    "JS symbols cannot be represented as a serde_json::Value".to_owned(),
+                ))
+            }
+            ValueType::External => {
+                return Err(Error::new(
+                    Status::InvalidArg,
+                    "External JS objects cannot be represented as a serde_json::Value".to_owned(),
+                ))
+            }
+            _ => {
+                return Err(Error::new(
+                    Status::InvalidArg,
+                    "Unknown JS variables cannot be represented as a serde_json::Value".to_owned(),
+                ))
+            }
+        });
+
+        Ok(val)
+    }
+}
+
+impl ToNapiValue for PVals {
+    unsafe fn to_napi_value(raw_env: sys::napi_env, val: Self) -> Result<sys::napi_value> {
+        let env = Env::from_raw(raw_env);
+        let mut obj = env.create_object()?;
+
+        for (k, v) in val.0.into_iter() {
+            obj.set(k, PVal(v))?;
+        }
+
+        unsafe { JsObject::to_napi_value(raw_env, obj) }
+    }
+}
+
+impl ToNapiValue for &PVals {
+    unsafe fn to_napi_value(raw_env: sys::napi_env, val: Self) -> Result<sys::napi_value> {
+        let env = Env::from_raw(raw_env);
+        let mut obj = env.create_object()?;
+
+        for (k, v) in val.0.iter() {
+            // Safety: the memory representation is the same for &PValue and &PVal.
+            let pv = unsafe { std::mem::transmute::<&PValue, &PVal>(v) };
+            obj.set(k, pv)?;
+        }
+
+        unsafe { JsObject::to_napi_value(raw_env, obj) }
+    }
+}
+
+impl FromNapiValue for PVals {
+    unsafe fn from_napi_value(env: sys::napi_env, napi_val: sys::napi_value) -> Result<Self> {
```

Comment:
this code panics if the value is `undefined`. I had similar problems when parsing the error details, and had to parse it to JsUnknown first and then match on the type to check that it was an JsObject before coercing into that.